### PR TITLE
Support batch jsonrpc http requests

### PR DESCRIPTION
**Pull request by dankostiuk**
# Description

Added support to handle batch JSON-RPC requests to align more closely with JSON-RPC support of geth (https://ethereum.org/en/developers/docs/apis/json-rpc/#json-rpc-support), iterating and dispatching each request in sequence for now. Also added more logging to log the request and response bodies.

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Checklist

- [ ] I have assigned this PR to myself
- [ ] I have added at least 1 reviewer
- [x] I have tested this code
- [ ] I have updated the README and other relevant documents (guides...)
- [x] I have added sufficient documentation both in code, as well as in the READMEs

### CODE REVIEW

File: jsonrpc/dispatcher_test.go [hunk 1]
```diff
@@ -153,3 +160,22 @@ func TestDispatcherFuncDecode(t *testing.T) {
 		}
 	}
 }
+
+func TestDispatcherBatchRequest(t *testing.T) {
+	s := newDispatcher(hclog.NewNullLogger(), newMockStore(), 0)
+	s.registerEndpoints()
+
+	resp, err := s.Handle([]byte(`[
+    {"id":1,"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["0x1", true]},
```
Comment (lazartravica) [inline jsonrpc/dispatcher_test.go:171]: Please add a` \t\r\n` and a whitespace to the beginning of the test case byte slice to make sure that we properly strip as per https://www.ietf.org/rfc/rfc4627.txt

**Review by lazartravica — CHANGES_REQUESTED** [review]
Congratulations on your first PR! I'd just kindly ask you to extend the tests a bit to cover the trimming functionality and that's it!

File: jsonrpc/dispatcher_test.go (hunk 1 above)
Comment (dankostiuk) [reply jsonrpc/dispatcher_test.go:171]: Done. Thanks @lazartravica

File: jsonrpc/jsonrpc.go [hunk 2]
```diff
@@ -210,10 +211,19 @@ func (j *JSONRPC) handle(w http.ResponseWriter, req *http.Request) {
 		handleErr(err)
 		return
 	}
+
+	// log request
+	req.Body = ioutil.NopCloser(bytes.NewReader(data))
+	buf := new(bytes.Buffer)
+	buf.ReadFrom(req.Body)
+	newStr := buf.String()
+	j.logger.Debug("handle", "request", newStr)
```
Comment (Vuksan) [inline jsonrpc/jsonrpc.go]: Why not just this?
```suggestion
	j.logger.Debug("handle", "request", string(data))
```

File: jsonrpc/dispatcher.go [hunk 3]
```diff
@@ -202,11 +203,58 @@ func (d *Dispatcher) HandleWs(reqBody []byte, conn wsConn) ([]byte, error) {
...
+	// handle batch requests
+	var requests []Request
+	if err := json.Unmarshal(reqBody, &requests); err != nil {
+		return nil, d.internalError("batch method", err)
+	}
+	var responses []Response
+	for _, req := range requests {
+		var response, err = d.handleReq(req)
+		if err != nil {
+			d.logger.Debug("failed to dispatch", "method", "batch method", "err", err)
```
Comment (Vuksan) [inline jsonrpc/dispatcher.go]: Call `internalError` instead?
```suggestion
			d.internalError("batch method", err)
```

File: jsonrpc/dispatcher.go [hunk 4]
```diff
@@ -202,11 +203,58 @@ func (d *Dispatcher) HandleWs(reqBody []byte, conn wsConn) ([]byte, error) {
...
+				Error: internalError,
+			}
+			responses = append(responses, errorResponse)
+			continue
+		}
+	
+		// unmarshal response from handleReq so that we can re-marshal as batch responses
+		var resp Response
+		if err := json.Unmarshal(response, &resp); err != nil {
+			d.logger.Debug("failed to dispatch", "method", "batch method", "err", err)
```
Comment (Vuksan) [inline jsonrpc/dispatcher.go]: Call `internalError` instead?
```suggestion
			d.internalError("batch method", err)
```

**Review by Vuksan — COMMENTED** [review]
Thanks for helping out! 🙏 
Left a couple of questions

**Review by Kourin1996 — APPROVED** [review]
Looks good to me 👏 I tested it locally and it works completely. I'm grateful for your nice improvements.

**Review by lazartravica — APPROVED** [review]
💯 👍

**Review by Vuksan — APPROVED** [review]
🎉

**Pull request merged**